src/Features/Core/Portable/Diagnostics/WorkspaceAnalyzerOptions.cs

```diff
@@ -10,13 +14,35 @@ namespace Microsoft.CodeAnalysis.Diagnostics
     /// </summary>
     internal sealed class WorkspaceAnalyzerOptions : AnalyzerOptions
     {
-        public WorkspaceAnalyzerOptions(AnalyzerOptions options, Workspace workspace)
+        private readonly Workspace _workspace;
+        private readonly OptionSet _optionSet;
+
+        public WorkspaceAnalyzerOptions(AnalyzerOptions options, OptionSet optionSet, Workspace workspace)
             : base(options.AdditionalFiles)
         {
-            this.Workspace = workspace;
+            _workspace = workspace;
+            _optionSet = optionSet;
         }
 
-        public Workspace Workspace { get; }
+        public HostWorkspaceServices Services => _workspace.Services;
+
+        public async Task<OptionSet> GetDocumentOptionSetAsync(SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var documentId = _workspace.CurrentSolution.GetDocumentId(syntaxTree);
```

Comment:
So my concern is this.  You're passing a Workspace + OptionSet, but then using Workspace.CurrentSolution instead of the actual Solution that we started all the work against.  That just seems busted (regardless of if this is what we already do).  I would just grab hte Solution snapshot.